src/unix/linux_like/android/b32/x86/mod.rs

```diff
@@ -574,6 +574,40 @@ pub const REG_EFL: ::c_int = 16;
 pub const REG_UESP: ::c_int = 17;
 pub const REG_SS: ::c_int = 18;
 
+// socketcall values from linux/net.h (only the needed ones, and not public)
+const SYS_ACCEPT4: ::c_int = 18;
+
+f! {
+    // Sadly, Android before 5.0 (API level 21), the accept4 syscall is not
+    // exposed by the libc. As work-around, we implement it as raw syscall.
+    // Note that for x86, the `accept4` syscall is not available either,
+    // and we must use the `socketcall` syscall instead.
+    // This workaround can be removed if the minimum Android version is bumped.
+    // When the workaround is removed, `accept4` can be moved back
+    // to `linux_like/mod.rs`
+    pub fn accept4(
+        fd: ::c_int,
+        addr: *mut ::sockaddr,
+        len: *mut ::socklen_t,
+        flg: ::c_int
+    ) -> ::c_int {
+        // Arguments are passed as array of `long int`
+        // (which is big enough on x86 for a pointer).
+        let mut args = [
+            fd as ::c_long,
+            addr as ::c_long,
+            len as ::c_long,
+            flg as ::c_long,
+        ];
+        socketcall(SYS_ACCEPT4, args[..].as_mut_ptr())
+    }
+}
+
+extern "C" {
+    // used to implement `accept4`, but not made public
+    fn socketcall(call: ::c_int, args: *mut ::c_long) -> ::c_int;
```

Comment:
I guess this is actually `__socketcall` on upstream?

Reply:
It seems there is no wrapper function for `socketcall` at all. Kinda makes sense for a syscall that is only an implementation detail for libc implementors. Looking back, including a Linux header to try and get a libc function declared doesn't make much sense :)

I changed the patch to use `::syscall(SYS_socketcall, SYS_ACCEPT4, ...)`, reverted the change to the tests and squashed it all.